### DISCUSSION TITLE
Can't run the lua scripts on 2.10.0

### DISCUSSION THREAD

**Discussion by une-glace** [EdgeTX/edgetx#5534] [LUA]
Problem Description:
TS16S radio gets stuck on loading when running a Lua script and doesn't proceed further.

Details:
Radio Model: TS16S
Firmware Version: 2.10.0

Issue Description:
I am trying to run a very simple Lua script on my TS16S radio:
```lua
local function runScript(event)
    print("Hello, World!")
    return 0  
end
return { run = runScript }
```
After placing the script in the SCRIPTS/ or SCRIPTS/FUNCTIONS/directory on the SD card and executing it from the radio, the screen gets stuck on loading and doesn't move forward.
The script itself seems error-free, but it still doesn't execute properly.
Additionally, if the script has syntax errors, the radio correctly reports them, indicating that the Lua engine itself seems to be functioning.

Steps I've Tried:
Updated the firmware and checked the SD card file system to ensure there are no errors.
Verified that the script file is saved in UTF-8 encoding without a BOM.
Tried using different SD cards, but the issue persists.

Assistance Needed:
Has anyone encountered a similar issue?
Are there any other debugging methods or solutions I can try?
Could this be a compatibility issue with the firmware version?

Thx a lot!!

**Comment by pfeerick** [comment]
Are you sure it is stuck? From the looks of that script, it is not doing any manipulation of the screen, since `print("Hello, World!")` will go out the console, not screen. 

Try adding a [lcd.clear()](https://luadoc.edgetx.org/part_iii_-_opentx_lua_api_reference/lcd-functions-less-than-greater-than-luadoc-begin-lcd/clear) and [lcd.drawText()](https://luadoc.edgetx.org/part_iii_-_opentx_lua_api_reference/lcd-functions-less-than-greater-than-luadoc-begin-lcd/drawtext) and see how you go then ;)

e.g. 

```lua
local function runScript(event)
    -- print("Hello, World!")
    lcd.clear()
    lcd.drawText(20, 20, "Hello, World!")
    return 0  
end
return { run = runScript }
```

Please note, if you are working on a function script, [per the documentation](https://luadoc.edgetx.org/part_i_-_script_type_overview/function_scripts), function scripts DO NOT have access to the display ;) They are not the same as [one time scripts](https://luadoc.edgetx.org/part_i_-_script_type_overview/one-time_scripts), i.e. scripts that you run from tools/sd browser

**Comment by une-glace** [comment]
Thank you for your response!

How can I view the console output? Is it by reading the SD card on the computer and then viewing it in the computer's Lua interpreter?

This is one of my test files. What I really want to achieve is to simulate pressing the SA button once after running the Lua script. During this process, no additional information will be displayed on the screen. This way, when I set the loop logic switch, I can let the remote control continuously send signals to the receiver. I've found that 0, 1, and 2 represent different states of the buttons, but how can I find the ID for the SA button?：）

----- 原始邮件 -----
发件人: "Peter Feerick" ***@***.***>
收件人: "EdgeTX/edgetx" ***@***.***>
抄送: "une_glace" ***@***.***>, "Author" ***@***.***>
发送时间: 星期六, 2024年 9 月 14日 上午 10:58:21
主题: Re: [EdgeTX/edgetx] Can't run the lua scripts on 2.10.0 (Discussion #5534)

Are you sure it is stuck? From the looks of that script, it is not doing any manipulation of the screen, since `print("Hello, World!")` will go out the console, not screen. 

Try adding a [lcd.clear()](https://luadoc.edgetx.org/part_iii_-_opentx_lua_api_reference/lcd-functions-less-than-greater-than-luadoc-begin-lcd/clear) and [lcd.drawText()](https://luadoc.edgetx.org/part_iii_-_opentx_lua_api_reference/lcd-functions-less-than-greater-than-luadoc-begin-lcd/drawtext) and see how you go then ;)

Please note, if you are working on a function script, [per the documentation](https://luadoc.edgetx.org/part_i_-_script_type_overview/function_scripts), function scripts DO NOT have access to the display ;) 


-- 
Reply to this email directly or view it on GitHub:
https://github.com/EdgeTX/edgetx/discussions/5534#discussioncomment-10643719

**Reply by pfeerick** [reply]
No, by either using the Companion Simulator, and viewing the log output while the script is running, or having the radio connected the PC, in VCP mode. 

> What I really want to achieve is to simulate pressing the SA button once after running the Lua script. 

Have you tried looking through the lua documentation at https://luadoc.edgetx.org yet?

If you're after the index, you probably want [getSwitchIndex()](https://luadoc.edgetx.org/part_iii_-_opentx_lua_api_reference/general-functions-less-than-greater-than-luadoc-begin-general/getswitchindex-positionname). Then you can use [getSwitchValue()](https://luadoc.edgetx.org/part_iii_-_opentx_lua_api_reference/general-functions-less-than-greater-than-luadoc-begin-general/getswitchvalue-switchindex) to see if the switch position is active.

**Reply by une-glace** [reply]
Thank you again for your response!

I would like to know if I need to link the library additionally in the Lua script when I want to use these general functions or model functions? Just like I need to configure specific libraries when calling mavros in C++. Or can I just call the function getSwitchIndex() directly? 
And sorry, I couldn't find the specific instructions in the documentation. I'm still not very familiar with lua. :(

Thank you for your reply! :D

**Reply by pfeerick** [reply]
No need to apologise, everyone gotta start somewhere, so just keep asking away ;) 

No need to include/import stuff, all the "builtin" stuff is just there.

**Reply by une-glace** [reply]
Thanks a lot! :D

I think I know how to code to my TX16S! But now I'm not in school, so I can't try these on the controller:( But two days later will be alright! :)

We are now celebrating the Mid-Autumn Festival, a time for family and friends to come together. I hope you can also spend a happy time with your loved ones!

Again, thanks for your reply! :)

**Reply by une-glace** [reply]
Hi! 
I successfully opened a simulator and ran the "Hello World" code. However, I can't find where the Lua console is, and I can't see the Lua script logs. Could you please tell me how I can view the output of functions like getSwitchIndex()? 
Thx a lot :)

**Reply by une-glace** [reply]
Good news！I found where I can see the log, but the result is not very good:(
I did some tests, and they went like this. I don't know where the problem is.
Below are my test results and the code:
![image](https://github.com/user-attachments/assets/2e55004a-6376-4c11-a27a-2b6192cdf965)
![image](https://github.com/user-attachments/assets/53bf8ffd-9bee-4373-b85d-ff8a926c2272)
Thanks for helping me:)
